signature/cose/cose.go

```diff
@@ -15,34 +26,524 @@ func init() {
+	if !ok || len(certs) == 0 {
+		return nil, &signature.MalformedSignatureError{Msg: "missing certificate chain"}
+	}
+	for _, certBytes := range certs {
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, err
+		}
+		certChain = append(certChain, cert)
+	}
+	signInfo.CertificateChain = certChain
+	// signingAgent
+	signInfo.UnsignedAttributes.SigningAgent = e.coseEnvelope.Headers.Unprotected[headerKeySigningAgent].(string)
+	// timestampSignature
+	signInfo.TimestampSignature = e.coseEnvelope.Headers.Unprotected[headerKeyTimeStampSignature].([]byte)
+
+	return &signInfo, nil
 }
 
+// NewEnvelope initializes an empty Cose signature envelope
 func NewEnvelope() signature.Envelope {
 	return &base.Envelope{
-		Envelope: &envelope{},
+		Envelope: &envelope{coseEnvelope: &cose.Sign1Message{}},
 	}
 }
 
+// ParseEnvelope parses envelopeBytes to a Cose signature envelope
 func ParseEnvelope(envelopeBytes []byte) (signature.Envelope, error) {
+	var coseEnvelope cose.Sign1Message
+	err := coseEnvelope.UnmarshalCBOR(envelopeBytes)
+	if err != nil {
+		return nil, err
+	}
 	return &base.Envelope{
-		Envelope: &envelope{},
+		Envelope: &envelope{coseEnvelope: &coseEnvelope},
 		Raw:      envelopeBytes,
 	}, nil
 }
+
+func validateCertificateChain(certChain []*x509.Certificate, signTime time.Time, expectedAlg cose.Algorithm, f func(string) error) error {
+	if len(certChain) == 0 {
+		return f("certificate-chain not present or is empty")
+	}
+
+	err := nx509.ValidateCodeSigningCertChain(certChain, signTime)
+	if err != nil {
+		return f(fmt.Sprintf("certificate-chain is invalid, %s", err))
+	}
+
+	resSignAlgo, err := getSignatureAlgorithm(certChain[0])
+	if err != nil {
+		return f(err.Error())
+	}
+	if resSignAlgo != expectedAlg {
+		return f("mismatch between signature algorithm derived from signing certificate and signing algorithm specified")
+	}
+
+	return nil
+}
+
+// getSignatureAlgorithm picks up a recommended signing algorithm for given certificate.
+func getSignatureAlgorithm(signingCert *x509.Certificate) (cose.Algorithm, error) {
+	keySpec, err := signature.ExtractKeySpec(signingCert)
+	if err != nil {
+		return 0, err
+	}
+	return getSignatureAlgorithmFromKeySpec(keySpec)
+}
+
+func getSignatureAlgorithmFromKeySpec(keySpec signature.KeySpec) (cose.Algorithm, error) {
+	switch keySpec.Type {
+	case signature.KeyTypeRSA:
+		switch keySpec.Size {
+		case 2048:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS256], nil
+		case 3072:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS384], nil
+		case 4096:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS512], nil
+		default:
+			return 0, errors.New("key size not supported")
+		}
+	case signature.KeyTypeEC:
+		switch keySpec.Size {
+		case 256:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES256], nil
+		case 384:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES384], nil
+		case 512:
```

Comment:
https://github.com/notaryproject/notation-core-go/blob/0553376c4e5d86681c177cbb5bfec37024dd69f6/signature/algorithm.go#L71-L80
Do we have a key size 512 if we use ecdsa?

Reply:
This part of code is from Binbin, which is already merged.